File: aoc22/2/main.py
from enum import Enum


f = open("2/input.txt")


class Tool(str, Enum):
    ROCK = "rock"
    PAPER = "paper"
    SCISSORS = "scissors"

    @staticmethod
    def from_str(char: str):
        if char in ["A", "X"]:
            return Tool.ROCK
        elif char in ["B", "Y"]:
            return Tool.PAPER
        elif char in ["C", "Z"]:
            return Tool.SCISSORS
        else:
            raise ValueError("Invalid tool code")


class Outcome(str, Enum):
    WIN = "Z"
    DRAW = "Y"
    LOSE = "X"


nemesis_map = {
    Tool.ROCK: Tool.PAPER,
    Tool.PAPER: Tool.SCISSORS,
    Tool.SCISSORS: Tool.ROCK,
}


def battle(p1: Tool, p2: Tool) -> int:
    if p1 == p2:
        return 0
    if p1 == nemesis_map[p2]:
        return -1
    else:
        return 1


def tool_score(tool: Tool):
    if tool == Tool.ROCK:
        return 1
    elif tool == Tool.PAPER:
        return 2
    else:
        return 3


def play(opponent: Tool, me: Tool):
    score = tool_score(me) + (battle(opponent, me) + 1) * 3
    return score


def find_tool(opponent: Tool, outcome: Outcome):
    if outcome == Outcome.DRAW:
        return opponent
    if outcome == Outcome.LOSE:
        return nemesis_map[nemesis_map[opponent]]
    else:
        return nemesis_map[opponent]


def play2(opponent: Tool, outcome: Outcome):
    my_tool = find_tool(opponent, outcome)
    score = tool_score(my_tool) + (battle(opponent, my_tool) + 1) * 3
    return score


scores = []
scores2 = []
for l in f.readlines():
    scores.append(play(Tool.from_str(l[0]), Tool.from_str(l[2])))
    scores2.append(play2(Tool.from_str(l[0]), Outcome(l[2])))

print(sum(scores))
print(sum(scores2))


f.close()

File: aoc22/2/src/main.rs
use std::env;
use std::fs;

#[derive(PartialEq)]
enum Tool {
    Rock,
    Paper,
    Scissors,
}
impl Tool {
    fn from_str(char: &str) -> Self{
        return match char {
                "A"|"X" => Tool::Rock,
                "B"|"Y" => Tool::Paper,
                _ => Tool::Scissors,
        };
    }

    fn score(tool: &Tool) -> i32{
        return match tool {
            Tool::Rock => 1,
            Tool::Paper => 2,
            Tool::Scissors => 3,
        };
    }
    fn nemesis_map(tool: &Tool) -> Tool {
        return match tool {
            Tool::Rock => Tool::Paper,
            Tool::Paper => Tool::Scissors,
            Tool::Scissors => Tool::Rock,
        }
    }
    fn battle(&self, p2: &Tool) -> i32 {
        if self == p2 {
            return 0;
        } 
        else if self == &Tool::nemesis_map(p2) {
            return -1; 
        }
        else{
            return 1;
        }
    }
}


#[derive(Debug)]
enum Outcome {
    Win,
    Draw,
    Lose,
}
impl Outcome {
    fn from_str(char: &str) -> Self{
        return match char {
                "X" => Outcome::Lose,
                "Y" => Outcome::Draw,
                _ => Outcome::Win,
        }
    }
    fn find_tool(&self, opponent: &Tool) -> Tool {
        return match self  {
            Outcome::Draw => match opponent {
                Tool::Paper => Tool::Paper,
                Tool::Rock => Tool::Rock,
                Tool::Scissors => Tool::Scissors,
            },
            Outcome::Lose => Tool::nemesis_map(&Tool::nemesis_map(opponent)),
            Outcome::Win => Tool::nemesis_map(opponent),
        }
    }
}


fn play(opponent: &Tool, me: &Tool) -> i32 {
    let score = Tool::score(me) + (opponent.battle(me) + 1)*3;
    return score;
}

fn play2(opponent: &Tool, outcome: &Outcome) -> i32 {
    let my_tool = outcome.find_tool(opponent);
    let score = Tool::score(&my_tool) + (opponent.battle(&my_tool) + 1)*3;
    return score;
}

fn main() {
    let mut current_dir = env::current_dir().unwrap().to_str().unwrap().to_owned();
    let today = "/2";
    if !current_dir.ends_with(today) {
        current_dir += today
    }
    let contents = fs::read_to_string("input.txt").expect("couldn't read file");
    let lines = contents.lines();
    let mut first: Vec<i32> = vec![];
    let mut second: Vec<i32> = vec![];

    for l in lines.into_iter() {
        let tool = Tool::from_str(l.chars().nth(0).unwrap().to_string().as_str());
        let second_tool = Tool::from_str(l.chars().nth(2).unwrap().to_string().as_str());
        let outcome = Outcome::from_str(l.chars().nth(2).unwrap().to_string().as_str());
        first.extend([play(&tool, &second_tool)]);
        second.extend([play2(&tool, &outcome)]);
    }
    let sum: i32 = first.iter().sum();
    let sum2: i32 = second.iter().sum();
    println!("{sum}");
    println!("{sum2}");
}
